core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java

```diff
@@ -223,7 +223,9 @@ public PartitionSpec apply() {
         // field IDs were not required for v1 and were assigned sequentially in each partition spec starting at 1,000.
         // to maintain consistent field ids across partition specs in v1 tables, any partition field that is removed
         // must be replaced with a null transform. null values are always allowed in partition data.
-        builder.add(field.sourceId(), field.fieldId(), field.name(), Transforms.alwaysNull());
+        // To avoid name conflict when add and remove same partition transform multiple times, field name will be
+        // replaced by field name append with field id.
+        builder.add(field.sourceId(), field.fieldId(), field.name() + "_" + field.fieldId(), Transforms.alwaysNull());
```

Comment:
Why do this instead of just renaming the field when a conflict is detected later? That seems like a better way to fix it to me.

Reply:
We have two choice when a a conflict is detected:
1. to rename the existing void transform, that will significant add complexity to [the check method](https://github.com/apache/iceberg/blob/35691728df3b6930045271b1e5663c8516b52301/api/src/main/java/org/apache/iceberg/PartitionSpec.java#L339), just to find which void transform conflict with the current added transform.
2. to rename the new transform, that will either broke default partition name consistency, or change the name assigned by user

Reply:
I don't think we want to modify that check method. Instead, this update class should be modified to rename void fields if necessary. That isn't very difficult because we can track whether there are existing fields with the same name and rename any field that uses a void transform.

Reply:
The newest commit modified update class to rename void fields if necessary. Please take a look again.